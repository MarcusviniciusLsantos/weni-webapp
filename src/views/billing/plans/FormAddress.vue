<template>
  <div class="billing-address-form">
    <div class="billing-address-form__duplicated">
      <UnnnicInput
        :label="$t('billing.address.cep')"
        v-model="$store.state.BillingSteps.billing_details.address.postal_code"
      />

      <UnnnicFormElement :label="$t('billing.address.country')">
        <UnnnicSelectSmart
          :value="
            [
              countries
                .map(({ native, iso2 }) => ({
                  value: iso2,
                  label: native,
                }))
                .find(
                  ({ value }) =>
                    value ===
                    $store.state.BillingSteps.billing_details.address.country,
                ),
            ].filter((i) => i)
          "
          @input="
            $store.state.BillingSteps.billing_details.address.country =
              $event[0].value
          "
          :options="
            [
              {
                value: '',
                label: $t('billing.address.select'),
              },
            ].concat(
              countries.map(({ native, iso2 }) => ({
                value: iso2,
                label: native,
              })),
            )
          "
          autocomplete
          autocompleteClearOnFocus
        >
        </UnnnicSelectSmart>
      </UnnnicFormElement>
    </div>

    <div class="billing-address-form__duplicated">
      <UnnnicFormElement
        v-if="statesOptions"
        :label="$t('billing.address.state')"
      >
        <UnnnicSelectSmart
          :value="
            [
              statesOptions
                .map((state) => ({
                  value: state,
                  label: state,
                }))
                .find(
                  ({ value }) =>
                    value ===
                    $store.state.BillingSteps.billing_details.address.state,
                ),
            ].filter((i) => i)
          "
          @input="
            $store.state.BillingSteps.billing_details.address.state =
              $event[0].value
          "
          :options="
            [
              {
                value: '',
                label: $t('billing.address.select'),
              },
            ].concat(
              statesOptions.map((state) => ({
                value: state,
                label: state,
              })),
            )
          "
          autocomplete
          autocompleteClearOnFocus
        >
        </UnnnicSelectSmart>
      </UnnnicFormElement>

      <UnnnicInput
        v-else
        :label="$t('billing.address.state')"
        :placeholder="$t('billing.address.type')"
        v-model="$store.state.BillingSteps.billing_details.address.state"
      />

      <UnnnicFormElement
        v-if="citiesOptions"
        :label="$t('billing.address.city')"
      >
        <UnnnicSelectSmart
          :value="
            [
              citiesOptions
                .map((city) => ({
                  value: city,
                  label: city,
                }))
                .find(
                  ({ value }) =>
                    value ===
                    $store.state.BillingSteps.billing_details.address.city,
                ),
            ].filter((i) => i)
          "
          @input="
            $store.state.BillingSteps.billing_details.address.city =
              $event[0].value
          "
          :options="
            [
              {
                value: '',
                label: $t('billing.address.select'),
              },
            ].concat(
              citiesOptions.map((city) => ({
                value: city,
                label: city,
              })),
            )
          "
          autocomplete
          autocompleteClearOnFocus
        >
        </UnnnicSelectSmart>
      </UnnnicFormElement>

      <UnnnicInput
        v-else
        :label="$t('billing.address.city')"
        :placeholder="
          isBrazilian && !brazilianStateSelected
            ? $t('billing.address.select_state')
            : $t('billing.address.type')
        "
        :disabled="isBrazilian && !brazilianStateSelected"
        v-model="$store.state.BillingSteps.billing_details.address.city"
      />
    </div>
    <UnnnicInput
      :label="$t('billing.address.address_title')"
      :placeholder="$t('billing.address.address_mask')"
      v-model="$store.state.BillingSteps.billing_details.address.line1"
    />
    <UnnnicInput
      :label="$t('billing.address.additional_info')"
      :placeholder="$t('billing.address.additional_info_mask')"
      v-model="$store.state.BillingSteps.billing_details.additionalInformation"
    />
  </div>
</template>

<script>
import { mapActions } from 'vuex';
import statesAndCitiesOfBrazil from '../../../assets/states-and-cities-of-brazil';
import countries from '../../../assets/countriesnames';

export default {
  name: 'BillingModal',

  props: {
    flow: String,
  },

  data() {
    return {
      countries,
    };
  },

  computed: {
    isBrazilian() {
      return (
        this.$store.state.BillingSteps.billing_details.address.country === 'BR'
      );
    },

    brazilianStateSelected() {
      if (this.isBrazilian) {
        return statesAndCitiesOfBrazil.estados.find(
          ({ nome }) =>
            nome ===
            this.$store.state.BillingSteps.billing_details.address.state,
        );
      }

      return null;
    },

    brazilianCitySelected() {
      if (this.brazilianStateSelected) {
        return this.brazilianStateSelected.cidades.find(
          (city) =>
            city ===
            this.$store.state.BillingSteps.billing_details.address.city,
        );
      }

      return null;
    },

    statesOptions() {
      if (this.isBrazilian) {
        return statesAndCitiesOfBrazil.estados.map(({ nome }) => nome);
      }

      return null;
    },

    citiesOptions() {
      if (this.brazilianStateSelected) {
        return this.brazilianStateSelected.cidades;
      }

      return null;
    },
  },

  watch: {
    '$store.state.BillingSteps.billing_details.address.country'() {
      if (this.isBrazilian) {
        if (this.brazilianStateSelected) {
          if (!this.brazilianCitySelected) {
            this.$store.state.BillingSteps.billing_details.address.city = '';
          }
        } else {
          this.$store.state.BillingSteps.billing_details.address.state = '';
        }
      }
    },

    '$store.state.BillingSteps.billing_details.address.state'() {
      if (this.isBrazilian) {
        this.$store.state.BillingSteps.billing_details.address.city = '';
      }
    },
  },

  methods: {
    ...mapActions(['setBillingStep']),
  },
};
</script>

<style lang="scss">
.billing-address-form {
  .unnnic-form,
  .unnnic-select {
    p {
      margin-top: 0;
    }
  }
  > .unnnic-form {
    margin-bottom: $unnnic-spacing-stack-md;
  }
  .weni-report {
    margin-top: $unnnic-spacing-stack-md;
  }

  &__duplicated {
    display: flex;
    margin-bottom: $unnnic-spacing-stack-md;
    > div {
      width: 100%;
      &:first-child {
        margin-right: $unnnic-spacing-stack-sm;
      }
    }
  }
}
</style>
