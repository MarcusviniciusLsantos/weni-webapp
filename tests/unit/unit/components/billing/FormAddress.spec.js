import { shallowMount, createLocalVue } from '@vue/test-utils';
import Router from 'vue-router';
import FormAddress from '@/views/billing/plans/FormAddress.vue';
import i18n from '@/utils/plugins/i18n';

import Vuex from 'vuex';
import { org } from '../../../__mocks__';
import statesAndCitiesOfBrazil from '@/assets/states-and-cities-of-brazil';
const localVue = createLocalVue();
localVue.use(Vuex);
localVue.use(Router);

const router = new Router();

describe('FormAddress.vue', () => {
  let wrapper;

  let actions;
  let store;
  let getters;
  let state;

  beforeEach(() => {
    actions = {};
    state = {
      BillingSteps: {
        billing_details: {
          address: {
            country: 'BR',
            state: 'AL',
            city: 'MCZ',
          },
        },
      },
    };
    getters = {
      currentOrg() {
        return org;
      },
    };
    store = new Vuex.Store({
      actions,
      getters,
      state,
    });
    wrapper = shallowMount(FormAddress, {
      localVue,
      i18n,
      router,
      propsData: {
        flow: 'test',
      },
      store,
      mocks: {
        $t: () => 'some specific text',
      },
      stubs: {
        unnnicToolTip: true,
        unnnicButton: true,
        unnnicSelect: true,
        unnnicInput: true,
      },
    });
  });

  it('renders a snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should redirect to org billing card', async () => {
    const spyRouter = jest.spyOn(wrapper.vm.$router, 'push');

    await wrapper.vm.back();

    expect(spyRouter).toHaveBeenCalledTimes(1);
  });

  // it('verify watch billing details address state', async () => {
  //   wrapper.vm.$store.state.BillingSteps.billing_details.address.country = 'BR';
  //   wrapper.vm.$store.state.BillingSteps.billing_details.address.state = 'test';

  //   expect(
  //     wrapper.vm.$store.state.BillingSteps.billing_details.address.city,
  //   ).toEqual('');
  // });

  describe('Computed isBrazilian', () => {
    it('should be truth', () => {
      wrapper.vm.$store.state.BillingSteps.billing_details.address.country =
        'BR';
      expect(wrapper.vm.isBrazilian).toBeTruthy();
    });
    it('should be falsy', () => {
      wrapper.vm.$store.state.BillingSteps.billing_details.address.country =
        'EUA';
      expect(wrapper.vm.isBrazilian).toBeFalsy();
    });
  });
  describe('Computed statesOptions', () => {
    it('should be null', () => {
      wrapper.vm.$store.state.BillingSteps.billing_details.address.country =
        'EUA';
      expect(wrapper.vm.statesOptions).toBeFalsy();
    });
    it('should be an array of states', () => {
      const states = statesAndCitiesOfBrazil.estados.map(({ nome }) => nome);
      wrapper.vm.$store.state.BillingSteps.billing_details.address.country =
        'BR';
      expect(wrapper.vm.statesOptions).toEqual(states);
    });
  });
  describe('Computed citiesOptions', () => {
    it('should be null', () => {
      expect(wrapper.vm.citiesOptions).toBeFalsy();
    });
  });
});
