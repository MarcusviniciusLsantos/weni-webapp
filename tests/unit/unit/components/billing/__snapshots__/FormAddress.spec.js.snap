// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FormAddress.vue > renders a snapshot 1`] = `
<div
  class="billing-address-form"
>
  <div
    class="billing-address-form__duplicated"
  >
    <unnnicinput-stub
      label="some specific text"
    />
    <unnnicformelement-stub
      label="some specific text"
    >
      <unnnicselectsmart-stub
        autocomplete=""
        autocompleteclearonfocus=""
        options="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]"
        value="[object Object]"
      />
    </unnnicformelement-stub>
  </div>
  <div
    class="billing-address-form__duplicated"
  >
    <unnnicformelement-stub
      label="some specific text"
    >
      <unnnicselectsmart-stub
        autocomplete=""
        autocompleteclearonfocus=""
        options="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]"
        value=""
      />
    </unnnicformelement-stub>
    <unnnicinput-stub
      disabled="true"
      label="some specific text"
      placeholder="some specific text"
      value="MCZ"
    />
  </div>
  <unnnicinput-stub
    label="some specific text"
    placeholder="some specific text"
  />
  <unnnicinput-stub
    label="some specific text"
    placeholder="some specific text"
  />
</div>
`;
